published/external/xdpapi_helper.h

```diff
@@ -0,0 +1,289 @@
+                              (XDP_API_CLIENT *)&_XdpApiContext, __VA_ARGS__)
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+NTSTATUS
+XdpHlpOpenApi(
+    _In_ UINT32 _XdpApiVersion,
+    _In_opt_ VOID *_ClientContext,
+    _In_opt_ XDP_API_ATTACH_FN *_ClientAttach,
+    _In_opt_ XDP_API_DETACH_FN *_ClientDetach,
+    _In_ const XDP_API_CLIENT_DISPATCH *_XdpApiClientDispatch,
+    _In_ const INT64 *_TimeoutMs
+    )
+{
+    return
+        XdpOpenApi(
+            _XdpApiVersion,
+            _ClientContext,
+            _ClientAttach,
+            _ClientDetach,
+            _XdpApiClientDispatch,
+            _TimeoutMs,
+            (XDP_API_CLIENT *)&_XdpApiContext,
+            (XDP_API_PROVIDER_DISPATCH **)&_XdpApi,
+            (XDP_API_PROVIDER_BINDING_CONTEXT **)&_ProviderBindingContext);
+}
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+VOID
+XdpHlpCloseApi(
+    VOID
+    )
+{
+    XdpUnloadApi((XDP_API_CLIENT *)&_XdpApiContext);
+}
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+VOID*
+XdpHlpGetRoutine(
+    _In_z_ const CHAR* RoutineName
+    )
+{
+    if (!ExAcquireRundownProtectionCacheAware(_XdpApiContext.RundownRef)) {
+        return NULL;
+    }
+    VOID* Routine = _XdpApi->XdpGetRoutine(RoutineName);
+    ExReleaseRundownProtectionCacheAware(_XdpApiContext.RundownRef);
+    return Routine;
+}
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+XDP_STATUS
+XdpHlpXskNotifyAsync2(
+    _In_ HANDLE Socket,
+    _In_ XSK_NOTIFY_FLAGS Flags,
+    _In_opt_ XSK_COMPLETION_CONTEXT CompletionContext,
+    _Out_ XSK_NOTIFY_RESULT_FLAGS *Result
+    )
+{
+    return _XdpApi->XskNotifyAsync2(Socket, Flags, CompletionContext, Result);
+}
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+VOID
+XdpHlpCloseHandle(
+    _In_ HANDLE Handle
+    )
+{
+    _XdpApi->XdpCloseHandle(Handle);
+}
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+XDP_STATUS
+XdpHlpXskCreate(
+    _In_opt_ PEPROCESS OwningProcess,
+    _In_opt_ PETHREAD OwningThread,
+    _In_opt_ PSECURITY_DESCRIPTOR SecurityDescriptor,
+    _Out_ HANDLE *Socket
+    )
+{
+    return
+        API_ABSTRACTION(
+            XskCreate,
+            OwningProcess,
+            OwningThread,
+            SecurityDescriptor,
+            Socket);
+}
+
+#else
+
+static const XDP_API_TABLE *_XdpApi;
+
+#define API_ABSTRACTION(FunctionCall, ...) _XdpApi->FunctionCall(__VA_ARGS__)
+
+
+inline
+HRESULT
+XdpHlpOpenApi(
+    _In_ UINT32 XdpApiVersion
+    )
+{
+    return XdpOpenApi(XdpApiVersion, &_XdpApi);
+}
+
+inline
+VOID
+XdpHlpCloseApi(
+    VOID
+    )
+{
+    XdpCloseApi(_XdpApi);
+}
+
+inline
+VOID*
+XdpHlpGetRoutine(
+    _In_z_ const CHAR* RoutineName
+    )
+{
+    return _XdpApi->XdpGetRoutine(RoutineName);
+}
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+XDP_STATUS
+XdpHlpXskCreate(
+    _Out_ HANDLE *Socket
+    )
+{
+    return _XdpApi->XskCreate(Socket);
+}
+
+#endif // defined(_KERNEL_MODE)
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+XDP_STATUS
+XdpHlpCreateProgram(
```

Comment:
do we really need all of these helpers?

Reply:
I don't know, but there will be no way to access all contexts 

```c
static const XDP_API_CLIENT _XdpApiContext = {0};
static const XDP_API_PROVIDER_DISPATCH *_XdpApi;
static const XDP_API_PROVIDER_BINDING_CONTEXT *_ProviderBindingContext;
```

Reply:
This can hide difference between user/kernel modes